nexus/tests/integration_tests/metrics_querier.rs

```diff
@@ -0,0 +1,415 @@
+use nexus_test_utils::ControlPlaneTestContext;
+use nexus_types::external_api::params;
+use nexus_types::external_api::views::OxqlQueryResult;
+use omicron_test_utils::dev::poll;
+use omicron_test_utils::dev::poll::wait_for_condition;
+use omicron_test_utils::dev::poll::CondCheckError;
+use oximeter::Datum;
+use oximeter::Measurement;
+use oximeter::TimeseriesSchema;
+use serde::de::DeserializeOwned;
+use slog::Logger;
+use std::borrow::Cow;
+use std::time::Duration;
+use uuid::Uuid;
+
+/// Error response expected for many of the condition closures passed to various
+/// methods of [`MetricsQuerier`] when the condition has not yet been satisfied.
+pub(super) struct MetricsNotYet {
+    note: Cow<'static, str>,
+}
+
+impl MetricsNotYet {
+    pub fn new<S>(note: S) -> Self
+    where
+        S: Into<Cow<'static, str>>,
+    {
+        Self { note: note.into() }
+    }
+}
+
+/// Helper for integration tests that want to interrogate metrics recorded into
+/// Clickhouse via Oximeter.
+///
+/// Many tests want to assert something of the form "after $MY_THING has
+/// started, what metrics have been recorded about it?" But this is a difficult
+/// question to ask without introducing flakiness, because there are multiple
+/// levels of asynchrony involved:
+///
+/// * Has the metric producer registered itself with Nexus yet?
+/// * Has Nexus told Oximeter about it yet?
+/// * Has Oximeter made a collection pass since the test did
+///   $ACTIVITY_OF_INTEREST?
+/// * Has the data collected by Oximeter been sent to Clickhouse yet?
+/// * If multinode Clickhouse is in play, has the data been replicated yet?
+///
+/// For the first two items, we have [`nexus_test_utils::wait_for_producer`].
+/// This type helps with the final three items.
+///
+/// Oximeter provides a `try_force_collect()` method. Historically, this would
+/// block until the collection was entirely complete (i.e., a new collection was
+/// made and all data collected had been inserted into single-node Clickhouse).
+/// With the move to multi-node Clickhouse, `try_force_collect()` is now "fire
+/// and forget": it instructs Oximeter to make a collection "soon", but does not
+/// wait for that collection to start nor for the data to be inserted.
+///
+/// `MetricsQuerier` provides a handful of methods that wrap
+/// [`wait_for_condition`] to allow tests to repeatedly issue metric queries
+/// until some condition is met. It's expected that tests will use a mix of
+/// returning an [`MetricsNotYet`] error (when the result indicates the metrics
+/// the test is waiting for haven't been inserted yet) and typical `assert!` /
+/// `panic!` tests (when the result indicates the data has been inserted but is
+/// somehow incorrect).
+pub(super) struct MetricsQuerier<'a, N> {
+    ctx: &'a ControlPlaneTestContext<N>,
+}
+
+impl<'a, N> MetricsQuerier<'a, N> {
+    const POLL_INTERVAL: Duration = Duration::from_secs(1);
+    const POLL_MAX: Duration = Duration::from_secs(60);
```

Comment:
Do we need any kind of configurability here?

Reply:
I'm not sure. I think "not right now"? With all the CI slowdowns we've seen I'd be a little hesitant to make this shorter, and it's hard to imagine an individual test needing something longer.